nomos-cli/src/cmds/chat/mod.rs

```diff
@@ -0,0 +1,260 @@
+            status_updates,
+            node: self.node.clone(),
+        };
+
+        run_app(&mut terminal, app);
+
+        // restore terminal
+        disable_raw_mode()?;
+        execute!(
+            terminal.backend_mut(),
+            LeaveAlternateScreen,
+            DisableMouseCapture
+        )?;
+        terminal.show_cursor()?;
+
+        Ok(())
+    }
+}
+
+fn run_app<B: Backend>(terminal: &mut Terminal<B>, mut app: App) {
+    let (message_tx, message_rx) = std::sync::mpsc::channel();
+    let node = app.node.clone();
+    std::thread::spawn(move || check_for_messages(message_tx, node));
+    loop {
+        terminal.draw(|f| ui::ui(f, &app)).unwrap();
+
+        if let Ok(update) = app.status_updates.try_recv() {
+            if let Status::Done = update {
+                app.message_in_flight = false;
+                app.message_status = None;
+            }
+            app.message_status = Some(update);
+            app.last_updated = Instant::now();
+        }
+
+        if let Ok(messages) = message_rx.try_recv() {
+            app.messages.extend(messages);
+        }
+
+        // Do not block rendering if there's no user input available
+        if !event::poll(Duration::from_millis(100)).unwrap() {
+            continue;
+        }
+
+        if let Event::Key(key) = event::read().unwrap() {
+            match key.code {
+                KeyCode::Enter => {
+                    if app.username.is_none() {
+                        app.username = Some(app.input.value().into());
+                    } else {
+                        // Do not allow more than one message in flight at a time for simplicity
+                        if !app.message_in_flight && !app.input.value().is_empty() {
+                            app.message_in_flight = true;
+                            app.payload_sender
+                                .send(
+                                    wire::serialize(&ChatMessage {
+                                        author: app.username.clone().unwrap(),
+                                        message: app.input.value().into(),
+                                    })
+                                    .unwrap()
+                                    .into(),
+                                )
+                                .unwrap();
+                        }
+                    }
+                    app.input.reset();
+                }
+                KeyCode::Esc => {
+                    return;
+                }
+                _ => {
+                    if !app.message_in_flight {
+                        app.input.handle_event(&Event::Key(key));
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+struct ChatMessage {
+    author: String,
+    message: String,
+}
+
+#[tokio::main]
+async fn check_for_messages(sender: Sender<Vec<ChatMessage>>, node: Url) {
+    let mut old_blocks = HashSet::new();
```

Comment:
How does this works as `tokio::main`? 😲

Reply:
`tokio::main` just automatically adds instrumentation to start a runtime when the function is called, does not really have to be the main function :)